gst/nnstreamer/tensor_source/tensor_src_iio.c

```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
+
+      id = g_ascii_strtoll (dir_entry->d_name + strlen (prefix), NULL, 10);
+      filename =
+          g_strdup_printf ("%s%s/%s", dir_name, dir_entry->d_name, NAME_FILE);
+
+      if (!g_file_get_contents (filename, &file_contents, NULL, &error)) {
+        GST_ERROR ("Unable to read %s, error: %s\n.", filename, error->message);
+        g_error_free (error);
+        goto error_free_filename;
+      }
+      g_free (filename);
+
+      if (!g_strcmp0 (file_contents, name)) {
+        ret = id;
+        g_free (file_contents);
+        break;
+      }
+      g_free (file_contents);
+    }
+  }
+
+  closedir (dptr);
+  return ret;
+
+error_free_filename:
+  g_free (filename);
+
+error_close_dir:
+  closedir (dptr);
+  return ret;
+}
+
+/**
+ * @brief get type info about the channel from the string
+ */
+static gboolean
+gst_tensor_src_iio_set_channel_type (GstTensorSrcIIOChannelProperties * prop,
+    const gchar * file_contents)
+{
+  gchar endianchar = '\0', signchar = '\0';
+  gint arguments_filled;
+  arguments_filled =
+      sscanf (file_contents, "%ce:%c%u/%u>>%u", &endianchar, &signchar,
+      &prop->mask_bits, &prop->storage_bits, &prop->shift);
+  if (arguments_filled < 5) {
+    return FALSE;
+  }
+
+  if (endianchar == 'b') {
+    prop->big_endian = TRUE;
+  } else {
+    prop->big_endian = FALSE;
+  }
+  if (signchar == 's') {
+    prop->is_signed = TRUE;
+  } else {
+    prop->is_signed = FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+ * @brief get generic name for channel from the string
+ *
+ * @return caller should free the returned string
+ */
+static gchar *
+gst_tensor_src_iio_get_generic_name (const gchar * channel_name)
+{
+  guint digit_len = 1;
+  gchar *generic_name;
+  guint channel_name_len = strlen (channel_name);
+
+  while (g_ascii_isdigit (channel_name[channel_name_len - digit_len])) {
+    digit_len++;
+  }
+  generic_name = g_strndup (channel_name, channel_name_len - digit_len + 1);
+
+  return generic_name;
+}
+
+/**
+ * @brief compare channels for sort based on their indices
+ */
+static gboolean
+gst_tensor_channel_list_sort_cmp (gconstpointer a, gconstpointer b)
+{
+  const GstTensorSrcIIOChannelProperties *a_ch = a;
+  const GstTensorSrcIIOChannelProperties *b_ch = b;
+  gint compare_result = a_ch->index - b_ch->index;
+  return compare_result;
+}
+
+/**
+ * @brief get info about all the channels in the device
+ * @return >=0 number of enabled channels
+ *         -1  if any error when scanning channels
+ */
+static guint
+gst_tensor_src_iio_get_all_channel_info (GstTensorSrcIIO * self,
+    const gchar * dir_name)
+{
+  DIR *dptr = NULL;
+  GError *error = NULL;
+  const struct dirent *dir_entry;
+  gchar *filename = NULL;
+  gchar *file_contents = NULL;
+  guint ret = -1;
+  guint value;
+  guint num_channels_enabled = 0;
+
+  if (!g_file_test (dir_name, G_FILE_TEST_IS_DIR)) {
+    GST_ERROR ("No channels available.");
+    return ret;
+  }
+
+  dptr = opendir (dir_name);
```

Comment:
Let's check if it's NULL or not. (it can pass g_file_test and fail opendir. e.g., permission error)

Reply:
Updated.